README
NAME
    URI::NamespaceMap - Class holding a collection of namespaces

VERSION
    Version 0.24

SYNOPSIS
      use URI::NamespaceMap;
      my $map = URI::NamespaceMap->new( { xsd => 'http://www.w3.org/2001/XMLSchema#' } );
      $map->namespace_uri('xsd')->as_string;
      my $foaf = URI::Namespace->new( 'http://xmlns.com/foaf/0.1/' );
      $map->add_mapping(foaf => $foaf);
      $map->add_mapping(rdf => 'http://www.w3.org/1999/02/22-rdf-syntax-ns#' );
      $map->list_prefixes;  #  ( 'foaf', 'rdf', 'xsd' )
      $map->foaf; # Returns URI::Namespace object

DESCRIPTION
    This module provides an object to manage multiple namespaces for creating
    URI::Namespace objects and for serializing.

METHODS
    `new ( [ \%namespaces | @prefixes | @uris ] )`
        Returns a new namespace map object. You can pass a hash reference with
        mappings from local names to namespace URIs (given as string or
        RDF::Trine::Node::Resource) or namespaces_map with a hashref. You may
        also pass an arrayref containing just prefixes and/or namespace URIs,
        and the module will try to guess the missing part. To use this
        feature, you need RDF::NS, XML::CommonNS or RDF::Prefixes, or
        preferably all of them.

    `add_mapping ( $name => $uri )`
        Adds a new namespace to the map. The namespace URI can be passed as
        string or a URI::Namespace object.

    `remove_mapping ( $name )`
        Removes a namespace from the map given a prefix.

    `namespace_uri ( $name )`
        Returns the URI::Namespace object (if any) associated with the given
        prefix.

    $name
        This module creates a method for all the prefixes, so you can say e.g.

          $map->foaf

        and get a URI::Namespace object for the FOAF namespace. Since
        URI::Namespace does the same for local names, you can then say e.g.

          $map->foaf->name

        to get a full URI.

    `list_namespaces`
        Returns an array of URI::Namespace objects with all the namespaces.

    `list_prefixes`
        Returns an array of prefixes.

    `uri ( $prefixed_name )`
        Returns a URI for an abbreviated string such as 'foaf:Person'.

    prefix_for `uri ($uri)`
        Returns the associated prefix (or potentially multiple prefixes, when
        called in list context) for the given URI.

    abbreviate `uri ($uri)`
        Complement to "namespace_uri". Returns the given URI in `foo:bar`
        format or `undef` if it wasn't matched, therefore the idiom

            my $str = $nsmap->abbreviate($uri_node) || $uri->as_string;

        may be useful for certain serialization tasks.

WARNING
    Avoid using the names 'can', 'isa', 'VERSION', and 'DOES' as namespace
    prefix, because these names are defined as method for every Perl object by
    default. The method names 'new' and 'uri' are also forbidden. Names of
    methods of Moose::Object must also be avoided.

AUTHORS
    Chris Prather, `<chris@prather.org>` Kjetil Kjernsmo, `<kjetilk@cpan.org>`
    Gregory Todd Williams, `<gwilliams@cpan.org>` Toby Inkster,
    `<tobyink@cpan.org>`

CONTRIBUTORS
    Dorian Taylor

BUGS
    Please report any bugs using github
    <https://github.com/kjetilk/URI-NamespaceMap/issues>

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc URI::NamespaceMap

COPYRIGHT & LICENSE
    Copyright 2012-2014 Gregory Todd Williams, Chris Prather and Kjetil
    Kjernsmo

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
URI-NamespaceMap
================

Created:      2012-08-27
Home page:    <https://metacpan.org/module/URI::NamespaceMap>
Bug tracker:  <https://github.com/kjetilk/URI-NamespaceMap/issues>
Maintainer:   Kjetil Kjernsmo <kjetilk@cpan.org>
Maintainer:   Gregory Todd Williams <gwilliams@cpan.org>
Maintainer:   Chris Prather (PERIGRIN) <chris@prather.org>

0.24	2014-12-10	Improve documentation.

 [ Bug Fixes ]
 - Use a string comparison in failing test.

 [ Other ]
 - Improve documentation, autoloader.
 - Reindent using Emacs smart-tabs.
 - Some code cleanup.

0.22	2014-10-15	Clean up source tree.

 - Clean up source tree.

0.21	2014-10-05	Fix missing Changes.

 [ Bug Fixes ]
 - Fix missing Changes.

0.20	2014-10-05	Moo refactoring.

 - Added: Warn the user when they have too old RDF::NS but not in automated
   tests.
 - Convert packaging to Dist::Inkt.
   Toby Inkster++
 - Port to Moo.
   Toby Inkster++

0.12	2014-09-29	Support IRI module.

 [ Bug Fixes ]
 - Silence warnings when URI doesn't have local part.

 [ Other ]
 - Added: Link Debian packages in metadata.
 - Added: Support the new IRI module.
   Gregory Todd Williams++
 - Added: Use Types::URI.
 - Added: Warn the user in test when they have too old RDF::NS.

0.10	2014-05-21	Compatibility with RDF::Trine::NamespaceMap.

 - Added: Make it possible to send local part to the uri method of
   URI::Namespace.
   Chris Prather++

0.08	2014-05-19	Prefix guessing.

 [ Bug Fixes ]
 - Fix breakage in terms with hyphens.
   Dorian Taylor++

 [ Other ]
 - Added: Prefix guessing using RDF::NS, RDF::Prefixes or XML::CommonNS.

0.06	2012-11-28	Bugfix release.

 [ Bug Fixes ]
 - URI::Namespace must be used from URI::NamespaceMap.

0.04	2012-10-23	Reverse lookups.

 - Added: Adding reverse lookups, i.e. abbreviation and prefix_for.
   Dorian Taylor++
 - Stop using blank node in makefile.ttl to fix version_from.

0.02	2012-08-30	Minor bugfixes.

0.01	2012-08-28	Initial release.


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "Namespace manipulation and prefix mapping for XML, RDF, etc.",
  "author"         => [
                        "Kjetil Kjernsmo <kjetilk\@cpan.org>",
                        "Gregory Todd Williams <gwilliams\@cpan.org>",
                        "Chris Prather (PERIGRIN) <chris\@prather.org>",
                      ],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.133380",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "URI-NamespaceMap",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          recommends => { "RDF::NS" => 20130802, "RDF::Prefixes" => 0, "XML::CommonNS" => 0 },
                          requires   => {
                                          "Carp" => 0,
                                          "IRI" => 0.004,
                                          "Module::Load::Conditional" => 0,
                                          "Moo" => "1.006000",
                                          "namespace::autoclean" => "0.20",
                                          "perl" => 5.008001,
                                          "Try::Tiny" => 0,
                                          "Type::Library" => "1.000000",
                                          "Types::URI" => 0.004,
                                          "URI" => 1.52,
                                        },
                        },
                        test => { requires => { "Test::Exception" => 0, "Test::More" => 0.88 } },
                      },
  "provides"       => {
                        "Types::Namespace"  => { file => "lib/Types/Namespace.pm", version => 0.12 },
                        "URI::Namespace"    => { file => "lib/URI/Namespace.pm", version => 0.12 },
                        "URI::NamespaceMap" => { file => "lib/URI/NamespaceMap.pm", version => 0.24 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "https://github.com/kjetilk/URI-NamespaceMap/issues" },
                        homepage     => "https://metacpan.org/module/URI::NamespaceMap",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/kjetilk/URI-NamespaceMap/",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/URI-NamespaceMap/project",
                      },
  "version"        => 0.24,
  "x_contributors" => [
                        "Dorian Taylor <dorian\@cpan.org>",
                        "Toby Inkster (TOBYINK) <tobyink\@cpan.org>",
                      ],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

